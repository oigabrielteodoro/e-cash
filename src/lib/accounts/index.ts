export const agencyNumberWithoutDigitMask = '9999'
export const accountNumberWithDigitMask = '999999-9'
