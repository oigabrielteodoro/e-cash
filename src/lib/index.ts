export * from './router'
export * from './formatters'
export * from './transformers'
export * from './input'
export * from './sidebar'
export * from './accounts'
export * from './masker'
export * from './matcher'
export * from './text'
