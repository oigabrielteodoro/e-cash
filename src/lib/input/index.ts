export * from './useInput'
export * from './getValidationErrors'
