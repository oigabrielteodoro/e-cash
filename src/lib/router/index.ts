export * from './Router'

export const SIGN_IN = '/'
export const DASHBOARD = '/dashboard'
