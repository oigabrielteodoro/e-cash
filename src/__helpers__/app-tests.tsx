import React, { useEffect, ReactElement, ReactNode } from 'react'
import { render as rtlRender } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import {
  DefaultOptions,
  QueryCache,
  QueryClient,
  QueryClientProvider,
} from 'react-query'
import { MemoryRouter, Routes, Route } from 'react-router-dom'
import { generateId } from 'lib'
import nock from 'nock'

import { pipe } from 'fp-ts/function'
import { TaskEither, map, mapLeft } from 'fp-ts/TaskEither'

import { AppLayout, NotificationContainer } from 'ui'
import { clearToken, queryConfigDefault, setToken } from 'client'

type Options = {
  initialRoute?: string
  initialRoutePath?: string
  authenticated?: boolean
  queryClient?: QueryClient
  routePaths?: string[]
}

type Props = {
  children?: ReactNode
  client?: QueryClient
}

type Callback = (result: unknown) => unknown
type MapAll = (
  fn: Callback,
) => (data: TaskEither<unknown, unknown>) => TaskEither<unknown, unknown>

const queryCache = new QueryCache()
const defaultQueryClient = new QueryClient({
  defaultOptions: queryConfigDefault as DefaultOptions,
  queryCache: queryCache,
})

export function ReactQueryWrapper({
  children,
  client = defaultQueryClient,
}: Props) {
  return <QueryClientProvider client={client}>{children}</QueryClientProvider>
}

export function ReactRouterWrapper({
  initialRoute = '/',
  children,
  routePaths,
  authenticated = true,
}: Options & Pick<Props, 'children'>) {
  function MockedPage({ title }: { title: string }) {
    if (authenticated) {
      return (
        <AppLayout>
          <AppLayout.Content>Page {title}</AppLayout.Content>
        </AppLayout>
      )
    }

    return <h1>Page {title}</h1>
  }

  return (
    <MemoryRouter initialEntries={[initialRoute]}>
      <Routes>
        <Route path={initialRoute} element={children} />
        {routePaths?.map((routePath) => (
          <Route
            key={routePath}
            path={routePath}
            element={<MockedPage title={routePath} />}
          />
        ))}
      </Routes>
    </MemoryRouter>
  )
}

export function ReactHookWrapper({
  initialRoute = '/',
  children,
  routePaths,
  authenticated = true,
  client = defaultQueryClient,
}: Options & Props) {
  return (
    <ReactQueryWrapper client={client}>
      <ReactRouterWrapper
        initialRoute={initialRoute}
        routePaths={routePaths}
        authenticated={authenticated}
      >
        {children}
      </ReactRouterWrapper>
    </ReactQueryWrapper>
  )
}

function render(
  ui: ReactElement,
  {
    initialRoute = '/',
    queryClient = defaultQueryClient,
    authenticated = true,
    routePaths,
  }: Options = {},
) {
  function Wrapper({ children }: Props) {
    useEffect(() => {
      if (authenticated) {
        setToken(generateId())

        return () => {
          clearToken()
        }
      }
    }, [])

    return (
      <ReactQueryWrapper client={queryClient}>
        <ReactRouterWrapper
          initialRoute={initialRoute}
          routePaths={routePaths}
          authenticated={authenticated}
        >
          {children}
        </ReactRouterWrapper>

        <NotificationContainer />
      </ReactQueryWrapper>
    )
  }

  return {
    ...rtlRender(ui, {
      wrapper: Wrapper,
    }),
  }
}

export function unsafe<T>(value: unknown) {
  return value as T
}

export const mapAll: MapAll = (fn) => (data) => {
  return pipe(data, map(fn), mapLeft(fn))
}

export function getErrorMessage(errors: unknown): string {
  return Array.isArray(errors) ? errors[0].message : ''
}

export * from '@testing-library/react'
export {
  render,
  userEvent,
  nock,
  defaultQueryClient as queryClient,
  queryCache,
}
