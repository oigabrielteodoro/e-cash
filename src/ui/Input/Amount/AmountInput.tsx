import React, {
  useState,
  forwardRef,
  InputHTMLAttributes,
  KeyboardEvent,
  ChangeEvent,
  FocusEvent,
  ForwardRefRenderFunction,
} from 'react'
import { FiDollarSign } from 'react-icons/fi'
import { AiOutlineExclamationCircle } from 'react-icons/ai'

import { pipe } from 'fp-ts/function'
import { fromNullable, map } from 'fp-ts/Option'

import { useInput } from 'hooks'
import { toDecimal } from 'lib'

import * as InputStyled from '../Input.styled'

export type AmountInputProps = {
  label: string
  error?: string
  variant?: 'primary' | 'secondary'
  defaultValue?: string
} & InputHTMLAttributes<HTMLInputElement>

function inputNumberParser(value?: string): string {
  return value ? value.replace(/\W/g, '') : ''
}

const ForwardAmountInput: ForwardRefRenderFunction<
  HTMLInputElement,
  AmountInputProps
> = (
  {
    id,
    name,
    variant = 'primary',
    label,
    defaultValue,
    error,
    onChange,
    onFocus,
    onBlur,
    ...rest
  },
  ref,
) => {
  const { isErrored, errorMessage } = useInput({
    defaultValue,
    error,
  })

  const [isFocused, setIsFocused] = useState(false)

  const [displayValue, setDisplayValue] = useState<string>(defaultValue ?? '')
  const formattedValue = displayValue
    ? toDecimal(inputNumberParser(displayValue))
    : undefined

  function handleOnKeyPress(event: KeyboardEvent<HTMLInputElement>) {
    if (!/[0-9]/.test(event.key)) event.preventDefault()
  }

  function handleOnChange(event: ChangeEvent<HTMLInputElement>) {
    pipe(
      event.currentTarget.value,
      fromNullable,
      map((amount) => {
        const parsedAmount = inputNumberParser(amount)

        setDisplayValue(parsedAmount)

        onChange &&
          onChange({
            ...event,
            currentTarget: {
              ...event.currentTarget,
              value: parsedAmount,
            },
          })
      }),
    )
  }

  function handleOnFocus(event: FocusEvent<HTMLInputElement>) {
    setIsFocused(true)

    onFocus && onFocus(event)
  }

  function handleOnBlur(event: FocusEvent<HTMLInputElement>) {
    setIsFocused(false)

    onBlur && onBlur(event)
  }

  return (
    <InputStyled.Wrapper>
      <label htmlFor={id}>{label}</label>
      <InputStyled.Container
        variant={variant}
        isFilled={false}
        isFocused={isFocused}
        isErrored={isErrored}
      >
        <input
          id={id}
          ref={ref}
          name={name}
          aria-label={name}
          value={formattedValue}
          defaultValue={defaultValue}
          onFocus={handleOnFocus}
          onBlur={handleOnBlur}
          onChange={handleOnChange}
          onKeyPress={handleOnKeyPress}
          {...rest}
        />
        <FiDollarSign size={22} />
      </InputStyled.Container>
      {error && (
        <InputStyled.ErrorContainer>
          <AiOutlineExclamationCircle size={18} />
          <span>{errorMessage}</span>
        </InputStyled.ErrorContainer>
      )}
    </InputStyled.Wrapper>
  )
}

export const AmountInput = forwardRef(ForwardAmountInput)
