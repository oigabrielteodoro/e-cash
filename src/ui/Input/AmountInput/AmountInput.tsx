import React, {
  useEffect,
  InputHTMLAttributes,
  KeyboardEvent,
  ChangeEvent,
} from 'react'
import { FiDollarSign } from 'react-icons/fi'
import { AiOutlineExclamationCircle } from 'react-icons/ai'
import { useFormContext } from 'react-hook-form'

import { useInput } from 'hooks'
import { toDecimal } from 'lib'

import * as InputStyled from '../Input.styled'

export type AmountInputProps = {
  name: string
  label: string
  error?: string
  variant?: 'primary' | 'secondary'
  defaultValue?: string
} & InputHTMLAttributes<HTMLInputElement>

function inputNumberParser(value?: string): string {
  return value ? value.replace(/\W/g, '') : ''
}

export function AmountInput({
  id,
  name,
  variant = 'primary',
  label,
  error,
  onFocus,
  onBlur,
  ...rest
}: AmountInputProps) {
  const { register, watch, setValue } = useFormContext()

  const value = watch(name)

  const {
    isFocused,
    isFilled,
    isErrored,
    errorMessage,
    handleOnBlur,
    handleOnFocus,
  } = useInput({
    defaultValue: value,
    error,
    onBlur,
    onFocus,
  })

  useEffect(() => {
    register(name)
  }, [name, register])

  const formattedValue = value ? toDecimal(inputNumberParser(value)) : ''

  function handleOnKeyPress(event: KeyboardEvent<HTMLInputElement>) {
    if (!/[0-9]/.test(event.key)) event.preventDefault()
  }

  function handleOnChange(event: ChangeEvent<HTMLInputElement>) {
    const value = event.currentTarget.value
    const amount = inputNumberParser(value)

    setValue(name, amount)
  }

  return (
    <InputStyled.Wrapper>
      <label htmlFor={id}>{label}</label>
      <InputStyled.Container
        variant={variant}
        isFilled={isFilled}
        isFocused={isFocused}
        isErrored={isErrored}
      >
        <input
          id={id}
          name={name}
          aria-label={name}
          value={formattedValue}
          onBlur={handleOnBlur}
          onFocus={handleOnFocus}
          onKeyPress={handleOnKeyPress}
          {...rest}
          onChange={handleOnChange}
        />
        <FiDollarSign size={22} />
      </InputStyled.Container>
      {error && (
        <InputStyled.ErrorContainer>
          <AiOutlineExclamationCircle size={18} />
          <span>{errorMessage}</span>
        </InputStyled.ErrorContainer>
      )}
    </InputStyled.Wrapper>
  )
}
