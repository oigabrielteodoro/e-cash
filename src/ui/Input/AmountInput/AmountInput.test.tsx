import React from 'react'
import { useForm } from 'react-hook-form'
import Faker from '@faker-js/faker'

import { render, screen, userEvent } from '__helpers__/app-tests'

import { toDecimal } from 'lib'
import { AmountInput } from './AmountInput'

type Props = {
  error?: string
}

function MockedComponent({ error }: Props) {
  const { register } = useForm()

  return <AmountInput label='Value' error={error} {...register('value')} />
}

describe('AmountInput', () => {
  it('should be able render correctly', () => {
    render(<MockedComponent />)

    expect(screen.getByRole('textbox')).toBeInTheDocument()
  })

  it('should be able type text correctly', async () => {
    const valueToType = Faker.datatype.number()

    render(<MockedComponent />)

    const input = screen.getByRole('textbox')

    expect(input).toBeInTheDocument()

    userEvent.type(input, valueToType.toString())

    expect(await screen.findByRole('textbox')).toHaveValue(
      toDecimal(valueToType),
    )
  })

  it('should be able render error message', async () => {
    render(<MockedComponent error='Value is required' />)

    expect(screen.getByRole('textbox')).toBeInTheDocument()
    expect(screen.getByText('Value is required')).toBeInTheDocument()
  })
})
