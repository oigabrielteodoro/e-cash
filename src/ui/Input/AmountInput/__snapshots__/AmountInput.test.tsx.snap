// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AmountInput should be able match snapshot 1`] = `
.c0 {
  border: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  position: relative;
}

.c0 label {
  position: absolute;
  font-size: 1.4rem;
  left: 1.2rem;
  top: -0.1rem;
  background: #ffffff;
  padding: 0 0.8rem;
  color: #a1a7ae;
  font-weight: 500;
  cursor: pointer;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: 0.1rem solid #a1a7ae;
  border-radius: 0.5rem;
  margin-top: 0.8rem;
  -webkit-transition: 300ms;
  transition: 300ms;
  width: 100%;
  background: #ffffff;
}

.c1 > svg,
.c1 .icon {
  margin-left: auto;
  margin-right: 2rem;
  color: #a1a7ae;
  -webkit-transition: 300ms;
  transition: 300ms;
}

.c1:hover {
  border: 0.1rem solid #3a67e1;
}

.c1:hover > svg,
.c1:hover .icon {
  color: #3a67e1;
}

.c1 input,
.c1 textarea {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  border: 0;
  background: transparent;
  padding: 1.8rem 2rem;
  color: #18171d;
  font-size: 1.6rem;
  border-radius: 0.5rem;
}

.c1 input::-webkit-input-placeholder,
.c1 textarea::-webkit-input-placeholder {
  color: #a1a7ae;
}

.c1 input::-moz-placeholder,
.c1 textarea::-moz-placeholder {
  color: #a1a7ae;
}

.c1 input:-ms-input-placeholder,
.c1 textarea:-ms-input-placeholder {
  color: #a1a7ae;
}

.c1 input::placeholder,
.c1 textarea::placeholder {
  color: #a1a7ae;
}

.c1 textarea {
  resize: none;
  padding: 0 2rem;
  margin: 2.4rem 0;
}

<div>
  <fieldset
    class="c0"
  >
    <label
      for="value"
    >
      Value
    </label>
    <div
      class="c1"
    >
      <input
        aria-label="value"
        id="value"
        name="value"
        value=""
      />
      <svg
        fill="none"
        height="22"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="22"
        xmlns="http://www.w3.org/2000/svg"
      >
        <line
          x1="12"
          x2="12"
          y1="1"
          y2="23"
        />
        <path
          d="M17 5H9.5a3.5 3.5 0 0 0 0 7h5a3.5 3.5 0 0 1 0 7H6"
        />
      </svg>
    </div>
  </fieldset>
</div>
`;
